pkg/storage/stores/indexshipper/downloads/index_set.go

```diff
@@ -0,0 +1,361 @@
+		// if we fail to open a boltdb file, lets skip it and let sync operation re-download the file from storage.
+		idx, err := t.openIndexFileFunc(fullPath)
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", fmt.Sprintf("failed to open existing index file %s, removing the file and continuing without it to let the sync operation catch up", fullPath), "err", err)
+			// Sometimes files get corrupted when the process gets killed in the middle of a download operation which causes boltdb client to panic.
+			// We already recover the panic but the lock on the file is not released by boltdb client which causes the reopening of the file to fail when the sync operation tries it.
+			// We want to remove the file failing to open to get rid of the lock.
+			if err := os.Remove(fullPath); err != nil {
+				level.Error(util_log.Logger).Log("msg", fmt.Sprintf("failed to remove index file %s which failed to open", fullPath))
+			}
+			continue
+		}
+
+		t.index[fileInfo.Name()] = idx
+	}
+
+	level.Debug(logger).Log("msg", fmt.Sprintf("opened %d local files, now starting sync operation", len(t.index)))
+
+	// sync the table to get new files and remove the deleted ones from storage.
+	err = t.sync(ctx, false)
+	if err != nil {
+		return
+	}
+
+	level.Debug(logger).Log("msg", "finished syncing files")
+
+	return
+}
+
+// Close Closes references to all the index.
+func (t *indexSet) Close() {
+	// stop the initialization if it is still ongoing.
+	t.cancelFunc()
+
+	err := t.indexMtx.lock(context.Background())
+	if err != nil {
+		level.Error(t.logger).Log("msg", "failed to acquire lock for closing index", "err", err)
+		return
+	}
+	defer t.indexMtx.unlock()
+
+	for name, db := range t.index {
+		if err := db.Close(); err != nil {
+			level.Error(t.logger).Log("msg", fmt.Sprintf("failed to close file %s", name), "err", err)
+		}
+	}
+
+	t.index = map[string]index.Index{}
+}
+
+func (t *indexSet) ForEach(ctx context.Context, callback index.ForEachIndexCallback) error {
+	if err := t.indexMtx.rLock(ctx); err != nil {
+		return err
+	}
+	defer t.indexMtx.rUnlock()
+
+	for _, idx := range t.index {
+		if err := callback(idx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DropAllDBs closes reference to all the open index and removes the local files.
+func (t *indexSet) DropAllDBs() error {
+	err := t.indexMtx.lock(context.Background())
+	if err != nil {
+		return err
+	}
+	defer t.indexMtx.unlock()
+
+	for fileName := range t.index {
+		if err := t.cleanupDB(fileName); err != nil {
+			return err
+		}
+	}
+
+	return os.RemoveAll(t.cacheLocation)
+}
+
+// Err returns the err which is usually set when there was any issue in Init.
+func (t *indexSet) Err() error {
+	return t.err
+}
+
+// LastUsedAt returns the time at which table was last used for querying.
+func (t *indexSet) LastUsedAt() time.Time {
+	return t.lastUsedAt
+}
+
+func (t *indexSet) UpdateLastUsedAt() {
+	t.lastUsedAt = time.Now()
+}
```

Comment:
Should these use the mutex, I saw an earlier comment that all `IndexSet` public functions are thread safe but it looks like we just wrap calls to this function within this file with mutex calls, so it wouldn't be thread safe. It's a relatively minor call, given that there's no computation happening here or in `LastUsedAt`.

Reply:
I didn't add a mutex for this because technically it is okay to concurrently overwrite a variable; it is just that we won't be sure which of the values would get assigned. Since we are not looking for a very precise last used time it is okay as long as the value gets updated. I do not have a strong opinion for this but happy to change it if you strongly feel we should add a mutex. What do you think?

Reply:
If, for the sake of usage of the stored value, it's not important that the value is 100% accurate, then I think not using a mutex is fine.